source/unix/openssl_rsa.c

```diff
@@ -0,0 +1,375 @@
+    return AWS_OP_SUCCESS;
+
+on_error:
+    EVP_PKEY_CTX_free(ctx);
+    return AWS_OP_ERR;
+}
+
+static int s_rsa_verify(
+    const struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_signature_algorithm algorithm,
+    struct aws_byte_cursor digest,
+    struct aws_byte_cursor signature) {
+    struct lc_rsa_key_pair *key_pair_impl = key_pair->impl;
+
+    EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new(key_pair_impl->key, NULL);
+    if (ctx == NULL) {
+        return aws_raise_error(AWS_ERROR_CAL_CRYPTO_OPERATION_FAILED);
+    }
+
+    if (s_reinterpret_evp_error_as_crt(
+        EVP_PKEY_verify_init(ctx), "EVP_PKEY_verify_init")) {
+        goto on_error;
+    }
+
+    if (s_set_signature_ctx_from_algo(ctx, algorithm)) {
+        goto on_error;
+    }
+
+    int error_code = EVP_PKEY_verify(ctx, signature.ptr, signature.len, digest.ptr, digest.len);
+    EVP_PKEY_CTX_free(ctx);
+
+    /*
+    * Verify errors slightly differently from the rest of evp functions. 0
+    * indicates signature does not pass verification, and negative indicates
+    * that something went wrong. Manually handle error instead of using helper.
+    */
+    if (error_code > 0) {
+        return AWS_OP_SUCCESS;
+    } else { 
+
+        uint32_t error = ERR_peek_error();
+        AWS_LOGF_ERROR(AWS_LS_CAL_RSA, "Calling function %s failed with %d and extended error %lu",
+            "EVP_PKEY_verify", error_code, (unsigned long)error);
+
+        if (error == 0) {
+            return aws_raise_error(AWS_ERROR_CAL_SIGNATURE_VALIDATION_FAILED);
+        } else if (error == -2) {
+            return aws_raise_error(AWS_ERROR_CAL_UNSUPPORTED_ALGORITHM);
+        } else {
+            return aws_raise_error(AWS_ERROR_CAL_CRYPTO_OPERATION_FAILED);
+        }
+    }
+
+on_error:
+    EVP_PKEY_CTX_free(ctx);
+    return AWS_OP_ERR;
+}
+
+static struct aws_rsa_key_vtable s_rsa_key_pair_vtable = {
+    .encrypt = s_rsa_encrypt,
+    .decrypt = s_rsa_decrypt,
+    .sign = s_rsa_sign,
+    .verify = s_rsa_verify,
+};
+
+struct aws_rsa_key_pair *aws_rsa_key_pair_new_from_private_key_pkcs1_impl(
+    struct aws_allocator *allocator,
+    struct aws_byte_cursor key) {
+    struct lc_rsa_key_pair *key_pair_impl = aws_mem_calloc(allocator, 1, sizeof(struct lc_rsa_key_pair));
+
+    aws_ref_count_init(&key_pair_impl->base.ref_count, &key_pair_impl->base, s_rsa_destroy_key);
+    key_pair_impl->base.impl = key_pair_impl;
+    key_pair_impl->base.allocator = allocator;
+    aws_byte_buf_init_copy_from_cursor(&key_pair_impl->base.priv, allocator, key);
+
+    RSA *rsa = NULL;
+    EVP_PKEY *private_key = NULL;
+
+    if (d2i_RSAPrivateKey(&rsa, (const uint8_t **)&key.ptr, key.len) == NULL) {
+        aws_raise_error(AWS_ERROR_CAL_CRYPTO_OPERATION_FAILED);
+        goto on_error;
+    }
+
+    private_key = EVP_PKEY_new();
+    if (private_key == NULL || EVP_PKEY_assign_RSA(private_key, rsa) == 0) {
+        aws_raise_error(AWS_ERROR_CAL_CRYPTO_OPERATION_FAILED);
+        goto on_error;
```

Comment:
I'm so confused about the ownership of `rsa`...

if `private_key == NULL`, does our error-handling here end up leaking `rsa`?

does EVP_PKEY_assign_RSA() transfer ownership of `rsa`, or just incref it somehow and we still need to free it ourselves?

OpenSSL APIs are terrible

Reply:
I asked BingGPT and it says that d2i_RSAPrivateKey() does allocate, so we do need to call `RSA_Free(rsa)` if `private_key == NULL`

and it says that EVP_PKEY_assign_RSA() does transfer ownership, so we should NOT call RSA_FREE() if assign_RSA() succeded

Looking at the openssl source code, it does look like an ownership transfer without any refcounting
https://github.com/openssl/openssl/blob/6246649d657127a031782b29ba6132a4203260b2/crypto/evp/p_lib.c#L408-L414

Reply:
yeah, you are right. added rsa free if either if the two conditions in the if fails.